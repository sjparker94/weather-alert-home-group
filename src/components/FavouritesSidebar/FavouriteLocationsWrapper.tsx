import styled from 'styled-components';

import { device } from '../../styles/breakpoint';

const FavouriteLocationsWrapper = styled.div`
    display: grid;
    grid-gap: ${props => props.theme.gutter};
    grid-template-columns: 1fr 1fr;
    padding-bottom: ${props => props.theme.gutter};
    @media ${device.laptopMMax} {
        display: block;
        @media ${device.laptopMMax} {
            padding-bottom: ${props => props.theme.gutterTablet};
        }
    }
`;

export default FavouriteLocationsWrapper;
