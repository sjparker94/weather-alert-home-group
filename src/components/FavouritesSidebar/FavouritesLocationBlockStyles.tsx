import styled from 'styled-components';
import { lighten, darken } from 'polished';
import { motion } from 'framer-motion';

import CircleButton from '../CircleButton/CircleButton';
import { device } from '../../styles/breakpoint';

const FavouritesLocationBlockStyles = styled(motion.div)`
    position: relative;
    margin: 0;
    @media ${device.laptopMMax} {
        margin-bottom: ${props => props.theme.gutterTablet};
    }
    > a {
        transition: all 0.25s ${props => props.theme.smoothAnimation};
        background-color: ${props => lighten(0.05, props.theme.favouritesSidebarBg())};
        border-radius: ${props => props.theme.mainBorderRadius};
        ${props => props.theme.lastItemMargin}
        box-shadow: ${props => props.theme.bs};

        &:hover,
        &:focus {
            transform: translateY(-4px) translateZ(0);
            background-color: ${props => lighten(0.1, props.theme.favouritesSidebarBg())};
        }
        &:hover {
            box-shadow: ${props => props.theme.hoverBs};
        }
        &:focus {
            box-shadow: ${props => props.theme.hoverBs}, inset 4px 0 0 ${props =>
    props.theme.primaryColor};

        }
        &:active {
            background-color: ${props => lighten(0.05, props.theme.favouritesSidebarBg())};
            transform: translateY(0px) translateZ(0);
            box-shadow: ${props => props.theme.contentBs};
        }
        &.selected {
            box-shadow: ${props => props.theme.bs}, inset 4px 0 0 ${props =>
    props.theme.primaryColor};
        }
    }
    /** Floating button */
    &:hover {
        ${CircleButton} {
            transition: all 0.25s ${props => props.theme.smoothAnimation};
            transform: translate(50%, calc(-50% - 2px)) translateZ(0);
            @media ${device.laptopMMax} {
                transform: translate(30%, calc(-30% - 2px)) translateZ(0);
            }
        }
    }
    ${CircleButton} {
        position: absolute;
        top: 0;
        right: 0;
        z-index: 1;
        transform: translate(50%, -50%);
        color: ${props => props.theme.textColor} !important;
        &:hover {
            transform: translate(50%, -50%) scale(1.1) translateZ(0);
        }
        &:active {
            transform: translate(50%, -50%) scale(0.95) translateZ(0);
        }
        @media ${device.laptopMMax} {
            transform: translate(30%, -30%);
            &:hover {
                transform: translate(30%, -30%) scale(1.1) translateZ(0);
            }
            &:active {
                transform: translate(30%, -30%) scale(0.95) translateZ(0);
            }
        }
    }

    > a {
        padding: ${props => props.theme.gutterPercentage(0.5)};
        display: flex;
    }
    .location-block-left {
        flex: 0 0 40px;
        margin-right: ${props => props.theme.gutterPercentage(0.25)};
    }
    .location-block-content {
        ${props => props.theme.lastItemMargin}
        flex: 1 1 100%;
    }
    .loading-message {
        color: ${props => props.theme.onDarkTextColorLight};
        /* ${props => props.theme.fontSize(12, 0)} */
    }
    .location-name {
        color: ${props => props.theme.onDarkTextColor};
        ${props => props.theme.headingFont('600')}
        ${props => props.theme.fontSize(18, 2)}
    }
    .weather-summary-item {
        color: ${props => props.theme.onDarkTextColor};
        ${props => props.theme.fontSize(24)}
        margin-bottom: 0;
        ${props => props.theme.font('400')}
        img,svg {
            vertical-align: middle;
            margin: 0 2px;
            opacity: 0.5;
        }
        sup,
        small {
            color: ${props => props.theme.onDarkTextColorLight};
            font-size: 0.7em;
        }
        sup {
            vertical-align: super;
        }
        .temp-icon {
            margin: 0 0 0 8px;
        }
    }
`;

export default FavouritesLocationBlockStyles;
