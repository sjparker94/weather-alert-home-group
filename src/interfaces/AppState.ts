import LocationsState from './LocationsState';
import SettingsState from './SettingsState';

interface AppState {
    locations: LocationsState;
    settings: SettingsState;
}

export default AppState;
