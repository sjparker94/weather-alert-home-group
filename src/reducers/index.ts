import { combineReducers } from 'redux';
import locations from './locationsReducer';
import settings from './settingsReducer';

const rootReducer = combineReducers({
    locations,
    settings,
});

export default rootReducer;
