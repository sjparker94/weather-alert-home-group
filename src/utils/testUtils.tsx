import Location from '../interfaces/Location';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from '../reducers/index';
import thunk from 'redux-thunk';
import theme from '../styles/theme';

export const fakeLocation = (): Location => ({
    coord: {
        lon: -0.13,
        lat: 51.51,
    },
    weather: [
        {
            id: 802,
            main: 'Clouds',
            description: 'scattered clouds',
            icon: '03d',
        },
    ],
    base: 'stations',
    main: {
        temp: 22.85,
        feels_like: 19.87,
        temp_min: 22.22,
        temp_max: 24.44,
        pressure: 1028,
        humidity: 31,
    },
    visibility: 10000,
    wind: {
        speed: 2.6,
        deg: 0,
    },
    clouds: {
        all: 40,
    },
    dt: 1594560101,
    sys: {
        type: 1,
        id: 1414,
        country: 'GB',
        sunrise: 1594526265,
        sunset: 1594584852,
    },
    timezone: 3600,
    id: 2643743,
    name: 'London',
    cod: 200,
});

const render = (ui: any, initialStore = {}, options = {}) => {
    const store = createStore(rootReducer, initialStore, applyMiddleware(thunk));

    const Providers = ({ children }: any) => (
        <Provider store={store}>
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
    );

    return rtlRender(ui, { wrapper: Providers, ...options });
};
// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
