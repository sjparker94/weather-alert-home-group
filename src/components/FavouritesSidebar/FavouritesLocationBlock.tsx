import React from 'react';
import { useDispatch } from 'react-redux';
import { NavLink } from 'react-router-dom';

import FavouritesLocationBlockStyles from './FavouritesLocationBlockStyles';
import CircleButton from '../CircleButton/CircleButton';
import { removeLocation } from '../../actions/locationsActions';
import Location from '../../interfaces/Location';
import { mpsToMph } from '../../utils/conversionUtils';
import windIcon from '../../assets/icons/wind-icon.svg';
import tempIcon from '../../assets/icons/temp-icon.svg';
import { BASE_IMAGE_URL } from '../../constants/siteInfo';

interface Props {
    locationData: Location;
    /** Passed in to show some details but wait until the latest information has been fetched to show weather info */
    isInitialLoad: boolean;
}

const FavouritesLocationBlock: React.FC<Props> = ({ locationData, isInitialLoad }) => {
    const dispatch = useDispatch();

    const {
        name,
        id,
        weather,
        wind,
        main: { temp },
    } = locationData;

    const deleteText = `Remove ${name} from Favourites`;
    const tempDisplay = Math.round(temp);
    const windSpeedDisplay = Math.round(mpsToMph(wind.speed));
    const isFahrenheit = false;

    const handleDeleteClick = () => {
        dispatch(removeLocation(id));
    };
    return (
        <FavouritesLocationBlockStyles
            initial={
                isInitialLoad
                    ? false
                    : {
                          scale: 0.5,
                          opacity: 0,
                          rotate: 30,
                      }
            }
            animate={{
                opacity: 1,
                scale: 1,
                rotate: 0,
            }}
            data-testid="favourite-location"
        >
            <NavLink
                to={`/location/${id}`}
                activeClassName="selected"
                data-testid="favourite-location-link"
            >
                <div className="location-block-left">
                    {!isInitialLoad && (
                        <img
                            src={`${BASE_IMAGE_URL}${weather[0].icon}.png`}
                            alt={weather[0].main}
                        />
                    )}
                </div>
                <div className="location-block-content">
                    <h3 className="location-name">{name}</h3>
                    {isInitialLoad ? (
                        <p className="loading-message">Fetching latest weather...</p>
                    ) : (
                        <>
                            <p className="weather-summary-item">
                                <img src={windIcon} width="30" height="30" alt="Wind speed" />
                                {windSpeedDisplay}
                                <small>mph</small>
                                <img
                                    src={tempIcon}
                                    width="26"
                                    height="26"
                                    alt="Temperature"
                                    className="temp-icon"
                                />
                                {tempDisplay}
                                <sup>{isFahrenheit ? '°F' : '°C'}</sup>
                            </p>
                        </>
                    )}
                </div>
            </NavLink>
            <CircleButton
                onClick={handleDeleteClick}
                title={deleteText}
                aria-label={deleteText}
                data-testid="favourite-delete-button"
                deleteButton
            >
                <span className="material-icons">delete</span>
            </CircleButton>
        </FavouritesLocationBlockStyles>
    );
};

export default FavouritesLocationBlock;
