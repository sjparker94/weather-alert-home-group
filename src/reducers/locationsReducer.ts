import produce, { Draft, castDraft } from 'immer';
import LocationsState from '../interfaces/LocationsState';
import neverReached from '../utils/neverReached';
import { AllActions } from '../actions/allActions';
import * as LocationsActionTypes from '../constants/actions';
import { findByProp } from '../utils/arrayUtils';
import Location from '../interfaces/Location';

// type LocationsActionTypes = any;
export const locationsInitialState: LocationsState = {
    data: [],
    isInitialLoad: true,
    getLocations: {
        isPending: false,
        success: false,
    },
    searchLocation: {
        isPending: false,
        success: false,
        error: null,
        data: null,
    },
};

const locations = produce((draft: Draft<LocationsState>, action: AllActions) => {
    switch (action.type) {
        case LocationsActionTypes.GET_LOCATIONS_REQUEST:
            draft.getLocations.isPending = true;
            break;
        case LocationsActionTypes.GET_LOCATIONS_SUCCESS:
            draft.data = castDraft(action.payload);
            draft.isInitialLoad = false;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_REQUEST:
            draft.searchLocation.isPending = true;
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_SUCCESS:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = null;
            draft.searchLocation.data = action.payload;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_CANCEL:
            draft.searchLocation.data = null;
            draft.searchLocation.error = null;
            break;
        case LocationsActionTypes.SEARCH_LOCATION_FAIL:
            draft.searchLocation.isPending = false;
            draft.searchLocation.error = action.payload;
            break;
        case LocationsActionTypes.ADD_LOCATION:
            // See if the location is already in favourites
            const currentIdx = draft.data.findIndex(findByProp<Location>('id', action.payload.id));

            if (currentIdx === -1) {
                draft.data.push(action.payload);
            } else {
                // Update existing location with the latest data fetched
                draft.data[currentIdx] = action.payload;
            }
            draft.searchLocation.data = null;
            break;
        default:
            neverReached(action);
    }
}, locationsInitialState);

export default locations;
