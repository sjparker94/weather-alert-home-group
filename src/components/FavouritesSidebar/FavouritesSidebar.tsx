import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import FavouritesSidebarStyles from './FavouritesSidebarStyles';
import FavouriteLocationsWrapper from './FavouriteLocationsWrapper';
import FavouritesLocationBlock from './FavouritesLocationBlock';
import useShallowEqualSelector from '../../hooks/useShallowEqualSelector';
import { getLocations } from '../../actions/locationsActions';
import Location from '../../interfaces/Location';
import AppState from '../../interfaces/AppState';

/** Persisted sidebar across all pages showing the users favourites */
const FavouritesSidebar: React.FC = () => {
    const locationsData = useShallowEqualSelector<Location[]>(state => state.locations.data);
    const isInitialLoad = useSelector<AppState, boolean>(state => state.locations.isInitialLoad);
    const dispatch = useDispatch();

    useEffect(() => {
        // On initial load we want the latest data so pass in items to go fetch it
        // Passing in it's own state data only to retrieve up to date items
        if (isInitialLoad) {
            dispatch(getLocations(locationsData));
        }
    }, [locationsData, isInitialLoad, dispatch]);

    return (
        <FavouritesSidebarStyles>
            <h2 className="main-title">
                <span className="material-icons star-icon">star</span> Your Favourites
            </h2>
            {locationsData.length ? (
                <FavouriteLocationsWrapper>
                    {locationsData.map(location => (
                        <FavouritesLocationBlock
                            key={location.id}
                            locationData={location}
                            isInitialLoad={isInitialLoad}
                        />
                    ))}
                </FavouriteLocationsWrapper>
            ) : (
                <div className="no-locations-saved" data-testid="favourites-no-locations">
                    <p>No locations have been selected use the form on the home page to add one</p>
                </div>
            )}
        </FavouritesSidebarStyles>
    );
};

export default FavouritesSidebar;
