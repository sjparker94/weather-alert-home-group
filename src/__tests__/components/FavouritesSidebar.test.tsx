import React from 'react';
import axios from 'axios';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter as Router } from 'react-router-dom';

// @ts-ignore
import { render, waitFor, fakeLocation, fireEvent, screen } from '../../utils/testUtils';
import FavouritesSidebar from '../../components/FavouritesSidebar/FavouritesSidebar';
import AppState from '../../interfaces/AppState';
import { locationsInitialState } from '../../reducers/locationsReducer';

jest.mock('axios');

const mockLocation = fakeLocation();
const mockLocation2 = fakeLocation({
    id: 123,
    name: 'Newcastle upon Tyne',
});

const mockInitalReduxStateWithData: AppState = {
    locations: {
        ...locationsInitialState,
        data: [mockLocation, mockLocation2],
    },
};

describe('<FavouritesSidebar />', () => {
    it('renders with no locations and matches snapshot', async () => {
        const { container } = render(
            <Router>
                <FavouritesSidebar />
            </Router>
        );

        // Check that the there are no favourites set
        expect(screen.queryAllByTestId('favourite-location')).toHaveLength(0);

        expect(container).toMatchSnapshot();
    });

    it('renders with 2 locations with fetching state before latest data is there', async () => {
        render(
            <Router>
                <FavouritesSidebar />
            </Router>,
            mockInitalReduxStateWithData
        );

        expect(screen.getAllByText(/fetching/i)).toHaveLength(2);
    });
    it('renders with 2 locations and matches snapshot', async () => {
        const { container } = render(
            <Router>
                <FavouritesSidebar />
            </Router>,
            mockInitalReduxStateWithData
        );

        // Wait for the latest data which displays the temperature of each location with icon
        const tempIcons = await screen.findAllByTitle('Temperature');
        const windSpeedIcons = await screen.findAllByTitle('Wind Speed');

        // Check that both of the mock loactions are in the dom and the weather is displaying
        expect(screen.getAllByTestId('favourite-location')).toHaveLength(2);
        expect(screen.getByText(mockLocation.name)).toBeInTheDocument();
        expect(screen.getByText(mockLocation2.name)).toBeInTheDocument();
        expect(tempIcons).toHaveLength(2);
        expect(windSpeedIcons).toHaveLength(2);

        expect(container).toMatchSnapshot();
    });

    it('removes location when the delete button is pressed', async () => {
        render(
            <Router>
                <FavouritesSidebar />
            </Router>,
            mockInitalReduxStateWithData
        );

        // Wait for the latest data which displays the temperature of each location with icon
        const deleteButtons = await screen.findAllByRole('button');

        fireEvent.click(deleteButtons[0]);

        const locations = screen.getAllByTestId('favourite-location');
        expect(deleteButtons[0]).not.toBeInTheDocument();
        expect(locations).toHaveLength(1);
    });

    it('it changes page/url when location link is clicked', async () => {
        render(
            <Router>
                <FavouritesSidebar />
            </Router>,
            mockInitalReduxStateWithData
        );

        // Wait for the latest data which displays the temperature of each location with icon
        const links = await screen.findAllByTestId('favourite-location-link');

        fireEvent.click(links[1]);
        expect(window.location.pathname).toBe('/location/123');
    });
});
