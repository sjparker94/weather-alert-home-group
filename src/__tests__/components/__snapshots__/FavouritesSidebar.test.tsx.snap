// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavouritesSidebar /> renders with 2 locations and matches snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm hUiZoz"
  >
    <h2
      class="main-title"
    >
      <span
        class="material-icons star-icon"
      >
        star
      </span>
       Your Favourites
    </h2>
    <div
      class="sc-AxirZ iqalEU"
    >
      <div
        class="sc-AxhCb leTKVX"
        data-testid="favourite-location"
        style="opacity: 1; transform: none;"
      >
        <a
          data-testid="favourite-location-link"
          href="/location/2643743"
        >
          <div
            class="location-block-left"
          >
            <img
              alt="Clouds"
              src="http://openweathermap.org/img/wn/03d.png"
            />
          </div>
          <div
            class="location-block-content"
          >
            <h3
              class="location-name"
            >
              London
            </h3>
            <p
              class="weather-summary-item"
            >
              <img
                alt="Wind speed"
                height="30"
                src="wind-icon.svg"
                width="30"
              />
              6
              <small>
                mph
              </small>
              <img
                alt="Temperature"
                class="temp-icon"
                height="26"
                src="temp-icon.svg"
                width="26"
              />
              23
              <sup>
                °C
              </sup>
            </p>
          </div>
        </a>
        <button
          aria-label="Remove London from Favourites"
          class="sc-AxiKw kpkgEV"
          data-testid="favourite-delete-button"
          title="Remove London from Favourites"
        >
          <span
            class="material-icons"
          >
            delete
          </span>
        </button>
      </div>
      <div
        class="sc-AxhCb leTKVX"
        data-testid="favourite-location"
        style="opacity: 1; transform: none;"
      >
        <a
          data-testid="favourite-location-link"
          href="/location/123"
        >
          <div
            class="location-block-left"
          >
            <img
              alt="Clouds"
              src="http://openweathermap.org/img/wn/03d.png"
            />
          </div>
          <div
            class="location-block-content"
          >
            <h3
              class="location-name"
            >
              Newcastle upon Tyne
            </h3>
            <p
              class="weather-summary-item"
            >
              <img
                alt="Wind speed"
                height="30"
                src="wind-icon.svg"
                width="30"
              />
              6
              <small>
                mph
              </small>
              <img
                alt="Temperature"
                class="temp-icon"
                height="26"
                src="temp-icon.svg"
                width="26"
              />
              23
              <sup>
                °C
              </sup>
            </p>
          </div>
        </a>
        <button
          aria-label="Remove Newcastle upon Tyne from Favourites"
          class="sc-AxiKw kpkgEV"
          data-testid="favourite-delete-button"
          title="Remove Newcastle upon Tyne from Favourites"
        >
          <span
            class="material-icons"
          >
            delete
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FavouritesSidebar /> renders with no locations and matches snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm hUiZoz"
  >
    <h2
      class="main-title"
    >
      <span
        class="material-icons star-icon"
      >
        star
      </span>
       Your Favourites
    </h2>
    <div
      class="no-locations-saved"
      data-testid="favourites-no-locations"
    >
      <p>
        No locations have been selected use the form on the home page to add one
      </p>
    </div>
  </div>
</div>
`;
